library/core/src/ptr/non_null.rs

```diff
@@ -1803,4 +1815,80 @@ mod verify {
         let maybe_null_ptr =  if kani::any() { xptr as *mut i32 } else { null_mut() };
         let _ = NonNull::new(maybe_null_ptr);
     }
+
+    // #[kani::proof_for_contract(NonNull::len)]
+    // pub fn non_null_check_len() {
+    //     const SIZE: usize = 20000;
+    //     // Create a non-deterministic array of size SIZE
+    //     let arr: [i8; SIZE] = kani::any();  
+    //     // Get a raw pointer to the array
+    //     let raw_ptr: *const i8 = arr.as_ptr();  
+    //     // Create a NonNull slice from the raw pointer
+    //     let non_null_slice: NonNull<[i8]> = unsafe { NonNull::slice_from_raw_parts(NonNull::new(raw_ptr as *mut i8).unwrap(), SIZE) };  
+
+    //     // Perform the length check
+    //     let len = non_null_slice.len();
+    //     assert!(len == SIZE);
+    // }
+
+    #[kani::proof_for_contract(NonNull::len)]
+    pub fn non_null_check_len() {
+        // Create a non-deterministic NonNull pointer using kani::any()
+        let non_null_ptr: NonNull<i8> = kani::any();
+        // Create a non-deterministic size using kani::any()
+        let size: usize = kani::any();
+        // Create a NonNull slice from the non-null pointer and size
+        let non_null_slice: NonNull<[i8]> = unsafe { NonNull::slice_from_raw_parts(non_null_ptr, size) };
+
+        // Perform the length check
+        let len = non_null_slice.len();
+        assert!(len == size);
+    }
+
+    #[kani::proof_for_contract(NonNull::is_empty)]
+    pub fn non_null_check_is_empty() {
+        const SIZE: usize = 0;
+        // Create a non-deterministic array of size SIZE
+        let arr: [i8; SIZE] = kani::any();  
+        // Get a raw pointer to the array
+        let raw_ptr: *const i8 = arr.as_ptr();  
+        // Create a NonNull slice from the raw pointer
+        let non_null_slice: NonNull<[i8]> = unsafe { NonNull::slice_from_raw_parts(NonNull::new(raw_ptr as *mut i8).unwrap(), SIZE) };  
+
+        // Perform the is_empty check
+        let is_empty = non_null_slice.is_empty();
+        assert!(is_empty);
+    }
+
+    #[kani::proof_for_contract(NonNull::is_aligned)]
+    pub fn non_null_slice_is_aligned_check() {
+        // Create a non-deterministic integer
+        let mut data: i32 = kani::any();  
+        // Get a raw pointer to the integer
+        let raw_ptr: *mut i32 = &mut data;
+        // NonNull pointer to the integer
+        let ptr = unsafe { NonNull::new(raw_ptr).unwrap() };
```

Comment:
I think it is more important at this point to have a harness that test as many cases as possible. I would recommend replacing this by an arbitrary `NonNull`.

Reply:
I'm not really sure what you mean by this. Can you maybe give an example?

Reply:
If you use the arbitrary implementation, now the pointer can have any value, including unaligned pointer, i.e.:

```rust
 let non_null_ptr: NonNull<i8> = kani::any();
```

The current harness creates a pointer from an aligned address, thus, it will only cover cases where this function returns true.